[quantization] Simplified nnq.Linear class

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24117 [quantization] JIT serialization for Conv2d
* #24116 [quantization] state_dict serialization for Conv2d + some bugfixes
* #24115 [quantization] Re-work Conv2d
* #24048 [quantization] JIT Serialization of nnq.Linear
* #24047 [quantization] State dict serialization of nnq.Linear
* **#24046 [quantization] Simplified nnq.Linear class**

`nnq.Linear` was a confusing mess of buffers/attributes and Tensor/not tensor members. This PR reworks it to consistently have only Python attributes, with the conversions handled explicitly by state_dict or __{get,set}state__ methods (added in PRs further up the stack

Differential Revision: [D16728345](https://our.internmc.facebook.com/intern/diff/D16728345)

Commit:
[quantization] Simplified nnq.Linear class

jerryzh168 reviewed (approved):
Looks great

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -4,6 +4,8 @@
 from ...modules.module import Module
 from ...modules.linear import Linear as NNLinear
 
+from typing import Optional
```
Does this work with Python 2?

jamesr66a replied:
It works as a backport library https://pypi.org/project/typing/

We use it in test_jit

Commit:
Update on "[quantization] Simplified nnq.Linear class"


`nnq.Linear` was a confusing mess of buffers/attributes and Tensor/not tensor members. This PR reworks it to consistently have only Python attributes, with the conversions handled explicitly by state_dict or __{get,set}state__ methods (added in PRs further up the stack

Commit:
Update on "[quantization] Simplified nnq.Linear class"


`nnq.Linear` was a confusing mess of buffers/attributes and Tensor/not tensor members. This PR reworks it to consistently have only Python attributes, with the conversions handled explicitly by state_dict or __{get,set}state__ methods (added in PRs further up the stack

jianyuh commented:
cc @dskhudia (See https://github.com/pytorch/pytorch/pull/24019 and https://github.com/pytorch/pytorch/pull/24059): Now the scale and zero_point are changed back to scalars, instead of a torch.tensor.

Commit:
Update on "[quantization] Simplified nnq.Linear class"


`nnq.Linear` was a confusing mess of buffers/attributes and Tensor/not tensor members. This PR reworks it to consistently have only Python attributes, with the conversions handled explicitly by state_dict or __{get,set}state__ methods (added in PRs further up the stack

Differential Revision: [D16728345](https://our.internmc.facebook.com/intern/diff/D16728345)

Commit:
Update on "[quantization] Simplified nnq.Linear class"


`nnq.Linear` was a confusing mess of buffers/attributes and Tensor/not tensor members. This PR reworks it to consistently have only Python attributes, with the conversions handled explicitly by state_dict or __{get,set}state__ methods (added in PRs further up the stack

Differential Revision: [D16728345](https://our.internmc.facebook.com/intern/diff/D16728345)
